source/common/upstream/original_dst_cluster.h

```diff
@@ -18,8 +18,21 @@
 namespace Envoy {
 namespace Upstream {
 
-using HostMapSharedPtr = std::shared_ptr<HostMap>;
-using HostMapConstSharedPtr = std::shared_ptr<const HostMap>;
+struct HostsForAddress {
+  HostsForAddress(HostSharedPtr& host) : host_(host), used_(true) {}
+
+  // Primary host for the address.
+  const HostSharedPtr host_;
+  // Hosts that are added concurrently with host_ are stored in this list.
+  std::vector<HostSharedPtr> hosts_;
```

Comment:
I'm still not convinced this is worth it versus just using a lock, given that we have to copy everything over to avoid using the lock? If we want to keep this can you add more comments on what concurrency is involved and how it happens?

Reply:
The lock is still there for the big table. The fine lock is just not necessary since it's fine for each worker to create their own host. We just need to keep track of all of them. Grabbing a lock on table entry has hard-to-measure effects on performance.